src/examples/ecore/ecore_promise2_example.c

```diff
@@ -0,0 +1,83 @@
+#include <Ecore.h>
+#include <Eina.h>
+#include <stdlib.h>
+
+static unsigned int _count = 0;
+
+static Eina_Bool
+_timeout(void *data)
+{
+   Efl_Promise2 *p = data;
+   Eina_Value *v = eina_value_new(EINA_VALUE_TYPE_INT);
+
+   eina_value_set(v, ++_count);
+   if (_count != 10)
+     efl_promise2_resolve(p, v);
+   else
+     efl_promise2_reject(p, v);
+   return EINA_FALSE;
+}
+
+Efl_Future2 *
+_async_future_get(void)
+{
+   Efl_Promise2 *p;
+
+   p = efl_promise2_new(NULL, NULL);
+   ecore_timer_add(0.1, _timeout, p);
+   return efl_future2_new(p, NULL, NULL, NULL, NULL);
+}
+
+static Eina_Value *
+_async_future_err(void *data EINA_UNUSED, const Eina_Value *value)
+{
+   unsigned int i;
+   eina_value_get(value, &i);
+   printf("Last step: %u\n", i);
+   ecore_main_loop_quit();
+   return NULL;
+}
+
+static Eina_Value *
+_async_future_end(void *data EINA_UNUSED, const Eina_Value *value)
+{
+   unsigned int i;
+   Eina_Value *v = eina_value_new(EINA_VALUE_TYPE_INT);
+
+   eina_value_get(value, &i);
+   printf("Step: %u\n", i);
+   efl_future2_then(_async_future_get(), _async_future_end,
+                    _async_future_err, NULL, NULL);
+   return NULL;
+}
+
+int
+main(int argc EINA_UNUSED, char *argv[] EINA_UNUSED)
+{
+   Efl_Future2 *f;
+
+   if (!eina_init())
+     {
+        fprintf(stderr, "Could not init eina\n");
+        return EXIT_FAILURE;
+     }
+
+   if (!ecore_init())
+     {
+        fprintf(stderr, "Could not init ecore\n");
+        goto err_ecore;
+     }
+
+   efl_future2_then(_async_future_get(), _async_future_end,
+                    _async_future_err, NULL, NULL);
+
+   ecore_main_loop_begin();
```

Comment:
it would be better to rewrite this as a set of examples, one for each "canonical case" below and also add unit tests covering these and also error cases (testing for error messages using eina_log infra).

Canonical cases, in JS notation. Suppose `img.load(path)` returns a future that propagates the `img` itself.

# Promise without futures

these should evaluate asynchronously (ie: do not cancel):

```js
img.load(path)
```

# Single "then" with asynchronous operation, no error check

```js
img.load(path)
   .then(i => console.log("loaded:", i.filename);
```

# Asynchronous operation with error check

```js
img.load(path)
   .catch(err => console.log("got error:", err);
```

# Already resolved values must evaluate from clean main loop context

That is, it's not dispatched immediately, will wait for next mainloop iteration:

```js
Promise.resolved(123)
   .then(v => console.log("printed third:", v);
console.log("printed first");

Promise.rejected("errmsg")
   .then(e => console.log("printed fourth:", e);
console.log("printed second");
```

# Chain of "then", changing values

```js
img.load(path)
   .then(i => i.size)
   .then(size => Math.sqrt(size[0] * size[1))
   .then(area => console.log("area is", area))
```

# Chain of "then", throwing errors

```js
img.load(path)
   .then(i => i.size)
   .then((size) => { throw new Error("exception!"); })
   .then(area => console.log("never executed", area))
   .catch(error => console.log("got error:", error);
```

# Chain of "then", with inner delayed operations

```js
img.load(path)
   .then((i) => {
      return new Promise((resolve, reject) => { // execution stops here, resumes on "resolve()"
         i.save(path + ".new")
            .then(i2 => resolve(i2)); // resumes "then" execution
      });
   }).then(i => console.log("load -> saved:", i.filename)); // prints ".new" version
```

# Chain of "then", with inner delayed operations and NO "then"

variation of the above, without the final "then". The "save" must execute asynchronously, but will not callback anything:

```js
img.load(path)
   .then((i) => {
      return new Promise((resolve, reject) => { // execution stops here, resumes on "resolve()"
         i.save(path + ".new")
            .then(i2 => resolve(i2)); // resumes "then" execution
      });
   }) // no "then", i.save() will happen async and will not callback anything
```

*MAYBE* https://github.com/promises-aplus/promises-tests may be of use, I didn't check the details.